### PR TITLE
add signature verification and unit tests for AWS plugin

### PR DESCRIPTION
**Pull request by crazytan**
**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/node_agent/na/aws.go [hunk 1]
```diff
@@ -41,16 +68,44 @@ func (na *awsPlatformImpl) IsProperPlatform() bool {
...
-// pluggable for different AWS deployments in the future.
+// This function should be pluggable for different AWS deployments.
 func (na *awsPlatformImpl) GetServiceIdentity() (string, error) {
 	return "", nil
 }
 
+func (na *awsPlatformImpl) GetInstanceIdentityDocument() ([]byte, error) {
+	cert, err := pki.ParsePemEncodedCertificate([]byte(AWSCertificatePem))
+	if err != nil {
+		return []byte{}, fmt.Errorf("Failed to parse AWS public certificate: %v", err)
```
Comment (wattli) [inline cmd/node_agent/na/aws.go]: what about return nil instead of []byte{} ?

Comment (wattli) [reply cmd/node_agent/na/aws.go]: same below

File: cmd/node_agent/na/aws.go [hunk 2]
```diff
@@ -15,12 +15,39 @@
...
+	"github.com/fullsailor/pkcs7"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials"
+	"istio.io/auth/pkg/pki"
+)
+
+const (
+	// AWSCertificatePem is the official public certificate for AWS
+	// copied from https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+	AWSCertificatePem = `-----BEGIN CERTIFICATE-----
```
Comment (myidpt) [inline cmd/node_agent/na/aws.go]: Could you put this content into a file like cert_aws.pem in testdata/? Thanks.

File: cmd/node_agent/na/aws.go [hunk 3]
```diff
@@ -41,16 +68,44 @@ func (na *awsPlatformImpl) IsProperPlatform() bool {
...
 }
 
-// Extract service identity from userdata. This function should be
-// pluggable for different AWS deployments in the future.
+// This function should be pluggable for different AWS deployments.
 func (na *awsPlatformImpl) GetServiceIdentity() (string, error) {
 	return "", nil
 }
 
+func (na *awsPlatformImpl) GetInstanceIdentityDocument() ([]byte, error) {
```
Comment (myidpt) [inline cmd/node_agent/na/aws.go]: Do you want to make this function private?

File: cmd/node_agent/na/aws_test.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+	server.Close()
+	if na.IsProperPlatform() {
+		t.Errorf("On Proper Platform: expected false")
+	}
+}
+
+func TestGetInstanceIdentityDocument(t *testing.T) {
+	server := initTestServer(
+		"/latest/dynamic/instance-identity/pkcs7",
```
Comment (myidpt) [inline cmd/node_agent/na/aws_test.go]: If possible, can you move this to a file in testdata/?

File: cmd/node_agent/na/aws.go (hunk 1 above)
Comment (crazytan) [reply cmd/node_agent/na/aws.go]: Done.

File: cmd/node_agent/na/aws.go (hunk 2 above)
Comment (crazytan) [reply cmd/node_agent/na/aws.go]: but it's not just used for test; it'll be used in production env to verify signature

File: cmd/node_agent/na/aws_test.go (hunk 4 above)
Comment (crazytan) [reply cmd/node_agent/na/aws_test.go]: will do

File: cmd/node_agent/na/aws.go (hunk 3 above)
Comment (crazytan) [reply cmd/node_agent/na/aws.go]: will do

File: cmd/node_agent/na/aws.go (hunk 2 above)
Comment (myidpt) [reply cmd/node_agent/na/aws.go]: You are right. Maybe put it in a file in a new dir data/?

Comment (crazytan) [reply cmd/node_agent/na/aws.go]: But wouldn't it require the data folder to be deployed alongside the Go binary for deployment?

Comment (myidpt) [reply cmd/node_agent/na/aws.go]: Yes, but will that be a problem?

Comment (myidpt) [reply cmd/node_agent/na/aws.go]: Anyway, we can change this later.

Comment (crazytan) [reply cmd/node_agent/na/aws.go]: My concern is that it kind of goes against the deployment model of Go (take a binary and deploy it any where). If putting the cert in the code is not desired, we can add a cmd option for this.
The deep issue is that AWS doesn't provide a way for us to dynamically retrieve the public cert, nor does it state whether the cert would be rotated in the future.

Comment (myidpt) [reply cmd/node_agent/na/aws.go]: Thanks for the explanation.

File: cmd/node_agent/na/aws.go (hunk 3 above)
Comment (crazytan) [reply cmd/node_agent/na/aws.go]: I'll change it in next PR

File: cmd/node_agent/na/aws_test.go (hunk 4 above)
Comment (crazytan) [reply cmd/node_agent/na/aws_test.go]: I'll change it in next PR

**Pull request merged**